Jenkinsfile

```diff
@@ -0,0 +1,164 @@
+            git checkout refs/remotes/origin/pr/${prNumber}/merge
+        """
+    }
+    return script
+}
+
+def mailBuildResults(String label, additionalEmails='') {
+    emailext (
+        subject: "[StanJenkins] ${label}: Job '${env.JOB_NAME} [${env.BUILD_NUMBER}]'",
+        body: """${label}: Job '${env.JOB_NAME} [${env.BUILD_NUMBER}]': Check console output at ${env.BUILD_URL}""",
+        recipientProviders: [[$class: 'RequesterRecipientProvider']],
+        to: "${env.CHANGE_AUTHOR_EMAIL}, ${additionalEmails}"
+    )
+}
+
+def runTests(String testPath) {
+    "runTests.py -j${env.PARALLEL} ${testPath} || echo ${testPath} failed"
+}
+
+def updateUpstream(String upstreamRepo) {
+    if (env.BRANCH_NAME == 'develop') {
+        sh "curl -O https://raw.githubusercontent.com/stan-dev/ci-scripts/master/jenkins/create-${upstreamRepo}-pull-request.sh"
+        sh "sh create-${upstreamRepo}-pull-request.sh"
+    }
+}
+
+pipeline {
+    agent none
+    parameters {
+        string(defaultValue: '', name: 'math_pr', description: "Leave blank "
+                + "unless testing against a specific math repo pull request, "
+                + "e.g. PR-640.")
+    }
+    options { skipDefaultCheckout() }
+    stages {
+        stage('Linting & Doc checks') {
+            agent any
+            steps {
+                script {
+                    retry(3) { checkout scm }
+                    sh setup(params.math_pr)
+                    stash 'StanSetup'
+                    parallel(
+                        CppLint: { sh "make cpplint" },
+                        documentation: { sh 'make doxygen' },
+                        manual: { sh 'make manual' },
+                        headers: { sh "make -j${env.PARALLEL} test-headers" },
+                        failFast: true
+                    )
+                }
+            }
+            post { always { deleteDir() } }
+        }
+        stage('Tests') {
+            failFast true
+            parallel {
+                stage('Windows Unit') {
+                    agent { label 'windows' }
+                    steps {
+                        unstash 'StanSetup'
+                        bat setupCC(false)
+                        bat runTests("src/test/unit")
+                        retry(2) { junit 'test/unit/**/*.xml' }
+                    }
+                    post { always { deleteDir() } }
+                }
+                stage('Windows Headers') { 
+                    agent { label 'windows' }
+                    steps {
+                        unstash 'StanSetup'
+                        bat setupCC()
+                        bat "make -j${env.PARALLEL} test-headers"
+                    }
+                    post { always { deleteDir() } }
+                }
+                stage('Unit') { 
+                    agent any
+                    steps {
+                        unstash 'StanSetup'
+                        sh setupCC(false)
```

Comment:
Why is the argument `false` here?

Reply:
Some tests will fail with `-Werror`, for example the multiple translation units test fails (at least on Jenkins) with:
```
[Unit] clang: error: test/unit/multiple_translation_units1.o: 'linker' input unused
[Unit] clang: error: test/unit/multiple_translation_units2.o: 'linker' input unused
```

(from http://d1m1s1b1.stat.columbia.edu:8080/job/Stan%20Pipeline/view/change-requests/job/PR-2414/23/console)

Though it doesn't fail like that for me locally...

Reply:
Makes sense. Perhaps that's using the wrong compiler? Seems like clang instead of clang++. But we can leave that for another day.

Reply:
The line above that error shows `clang++` was the thing actually called:
```
[Unit] clang++ -Werror -Wall -I . -isystem lib/stan_math/lib/eigen_3.3.3 -isystem lib/stan_math/lib/boost_1.64.0 -isystem lib/stan_math/lib/cvodes_2.9.0/include -std=c++1y -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION -Wno-unused-function -Wno-uninitialized -I src -isystem lib/stan_math/ -DFUSION_MAX_VECTOR_SIZE=12 -Wno-unused-local-typedefs -stdlib=libc++ -Wno-unknown-warning-option -Wno-tautological-compare -Wsign-compare -DNO_FPRINTF_OUTPUT -pipe  -c -shared -o test/unit/libmultiple.so test/unit/multiple_translation_units1.o test/unit/multiple_translation_units2.o
```

I am a bit confused about this, but these jobs didn't have `-Werror` in the existing Jenkins jobs either, so for now I think we're preserving status quo in this area.